doc/libpmem/pmem_memmove_persist.3.md

```diff
@@ -66,72 +70,108 @@ void *pmem_memset_nodrain(void *pmemdest, int c, size_t len);
-
-	pmem_persist(pmemdest, len);
+For example, the following code is functionally equivalent to **pmem_memmove**() (with flags equal to 0):
 
-	return retval;
-}
+```c
+	memmove(dest, src, len);
+	pmem_persist(dest, len);
 ```
 
-Calling **pmem_memmove_persist**() may out-perform the above code,
-however, since the **libpmem**(7) implementation may take advantage of the
-fact that *pmemdest* is persistent memory and use instructions such as
-*non-temporal* stores to avoid the need to flush processor caches.
+Calling **pmem_memmove**() may out-perform the above code, because
+**libpmem**(7) implementation may take advantage of the fact that *pmemdest*
+is persistent memory and use instructions such as *non-temporal* stores to
+avoid the need to flush processor caches.
 
 >WARNING:
 Using these functions where **pmem_is_pmem**(3) returns false
-may not do anything useful. Use the normal libc functions in that case.
+may not do anything useful. Use libc functions in that case.
 
+Unlike libc implementation, **libpmem** functions guarantee that if destination
+buffer address and length are 8 byte aligned then all stores will be performed
+using at least 8 byte store instructions. This means that a series of 8 byte
+stores followed by **pmem_persist**(3) can be safely replaced by a single call
+to one of the above functions.
 
-The **pmem_memmove_nodrain**(), **pmem_memcpy_nodrain**() and
-**pmem_memset_nodrain**() functions are similar to
-**pmem_memmove_persist**(), **pmem_memcpy_persist**(), and
-**pmem_memset_persist**() described above, except they skip the final
-**pmem_drain**() step. This allows applications to optimize cases where
-several ranges are being copied to persistent memory, followed by a
-single call to **pmem_drain**(). The following example illustrates how
-these functions might be used to avoid multiple calls to
-**pmem_drain**() when copying several ranges of memory to pmem:
+The *flags* argument of all of the above functions has the same meaning.
+It can be 0 or a bitwise OR of one or more of the following flags:
+
++ **PMEM_MEM_NODRAIN** - modifies the behavior to skip the final
+  **pmem_drain**() step. This allows applications to optimize cases where
+  several ranges are being copied to persistent memory, followed by a single
+  call to **pmem_drain**(). The following example illustrates how this flag
+  might be used to avoid multiple calls to **pmem_drain**() when copying several
+  ranges of memory to pmem:
 
 ```c
 /* ... write several ranges to pmem ... */
-pmem_memcpy_nodrain(pmemdest1, src1, len1);
-pmem_memcpy_nodrain(pmemdest2, src2, len2);
+pmem_memcpy(pmemdest1, src1, len1, PMEM_MEM_NODRAIN);
+pmem_memcpy(pmemdest2, src2, len2, PMEM_MEM_NODRAIN);
 
 /* ... */
 
 /* wait for any pmem stores to drain from HW buffers */
 pmem_drain();
 ```
 
+The remaining flags say *how* the operation should be done, and are merely hints.
+
++ **PMEM_MEM_NONTEMPORAL** - Use non-temporal instructions.
```

Comment:
At least ARM has non temporal stores (http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.den0024a/CJACGJJF.html), so these flags are not x86-specific.

I added both nontemporal and nocache, because the first name instructs how we should perform the copy exactly, while the second describes certain aspect of non-temporal instruction.

On x86 nontemporal means nocache+wc, however I can imagine an architecture where there is different method of bypassing caches without using non-temporal instructions (which might have some negative aspect on that architecture).

Does it make sense to you?

Reply:
Ok, after a bit of discussion we came to the conclusion that CACHE and NOCACHE flags are not needed, at least for now. I removed them.